### PR TITLE
Added support for via() relations in model generator

### PR DESCRIPTION
**Pull request by rhertogh**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes
| Tests pass?   | yes
| Fixed issues  | #14 , #212

### CODE REVIEW

**Comment by rhertogh** [comment]
Please note that that while the generated model is backwards compatible by default (viaTable is the default option), the signature of `Generator::checkJunctionTable()` has changed. This might cause problems when someone has extended the Generator and uses the checkJunctionTable function in custom code.
My suggestion is to add a `upgrade.md` to describe the change. Please let me know if you agree and I'll add it.

**Comment by schmunk42** [comment]
@yiisoft/core-developers Please do a `2.2.x` release if a method signature is changed.

**Comment by samdark** [comment]
@schmunk42 OK but I don't see any signature changes in the pull request.

**Comment by rhertogh** [comment]
@samdark 
>  OK but I don't see any signature changes in the pull request.
My comment was a bit unclear, it's not the signature of the function but the signature of the returned value (the array now has an extra dimension).

**Comment by rhertogh** [comment]
I've changed the changelog version in the pr  to 2.2.0 and added UPGRADE.md.
The Breaks BC? is changed to 'yes' in the pr opening post.

File: src/generators/model/Generator.php [hunk 1]
```diff
@@ -104,6 +111,7 @@ public function attributeLabels()
             'modelClass' => 'Model Class Name',
             'baseClass' => 'Base Class',
             'generateRelations' => 'Generate Relations',
+            'generateJunctionRelationMode' => 'Generate Junction Relation Mode',
```
Comment (samdark) [inline src/generators/model/Generator.php]:
```suggestion
            'generateJunctionRelationMode' => 'Generate Junction Relations As',
```

File: src/generators/model/Generator.php [hunk 2]
```diff
@@ -490,25 +501,65 @@ private function generateManyManyRelations($table, $fks, $relations)
...
-            $relationName = $this->generateRelationName($relations, $table0Schema, key($secondKey), true);
-            $relations[$table0Schema->fullName][$relationName] = [
-                "return \$this->hasMany($className1::className(), $link)->viaTable('"
-                . $this->generateTableName($table->name) . "', $viaLink);",
-                $className1,
-                true,
-            ];
+            $link = $this->generateRelationLink(array_flip($secondKey[0]));
+            $relationName = $this->generateRelationName($relations, $table0Schema, key($secondKey[0]), true);
+            if ($this->generateJunctionRelationMode == self::JUNCTION_RELATION_VIA_TABLE) {
```
Comment (samdark) [inline src/generators/model/Generator.php]:
```suggestion
            if ($this->generateJunctionRelationMode === self::JUNCTION_RELATION_VIA_TABLE) {
```

File: src/generators/model/Generator.php [hunk 3]
```diff
@@ -490,25 +501,65 @@ private function generateManyManyRelations($table, $fks, $relations)
...
+                $relations[$table0Schema->fullName][$relationName] = [
+                    "return \$this->hasMany($className1::className(), $link)->viaTable('"
+                    . $this->generateTableName($table->name) . "', " . $this->generateRelationLink($firstKey[0]) . ");",
+                    $className1,
+                    true,
+                ];
+            } elseif ($this->generateJunctionRelationMode == self::JUNCTION_RELATION_VIA_MODEL) {
```
Comment (samdark) [inline src/generators/model/Generator.php]:
```suggestion
            } elseif ($this->generateJunctionRelationMode === self::JUNCTION_RELATION_VIA_MODEL) {
```

File: src/generators/model/Generator.php [hunk 4]
```diff
@@ -490,25 +501,65 @@ private function generateManyManyRelations($table, $fks, $relations)
...
+                    throw new Exception('Foreign key for junction table not found.');
+                }
+                $relations[$table0Schema->fullName][$relationName] = [
+                    "return \$this->hasMany($className1::className(), $link)->via('"
+                    . lcfirst($foreignRelationName) . "');",
+                    $className1,
+                    true,
+                ];
+            } else {
+                throw new InvalidConfigException('Unknown generateViaRelationMode ' . $this->generateJunctionRelationMode);
```
Comment (samdark) [inline src/generators/model/Generator.php]: This check could be done once at the top of the method. Would be a bit less cognitive load.

Comment (rhertogh) [reply src/generators/model/Generator.php]: Check is moved to the top of the function. Using new JUNCTION_RELATION_MODES array to check against. This array is now also used in the `rules()`.

File: src/generators/model/Generator.php [hunk 5]
```diff
@@ -35,6 +35,11 @@ class Generator extends \yii\gii\Generator
     const JUNCTION_RELATION_VIA_TABLE = 'table';
     const JUNCTION_RELATION_VIA_MODEL = 'model';
 
+    const JUNCTION_RELATION_MODES = [
```
Comment (samdark) [inline src/generators/model/Generator.php]: That won't work for older PHP versions.

Comment (rhertogh) [reply src/generators/model/Generator.php]: dropped the array constant, now defining the array in functions where needed.

**Comment by samdark** [comment]
Thank you!

**Pull request merged**